### DISCUSSION TITLE
How to choose toolset for  library compilation ?

### DISCUSSION THREAD

**Discussion by kuzawskak** [microsoft/vcpkg#31858] [Q&A]
I have downloaded the newest version of vcpkg . How can I choose msvc version (I have installed several msvc versions on my computer) to compile?
I thought that "vcpkg.exe install apr --toolset=msvc-142" will work - but it doesn't. It doesn't recognize --toolset option... (error: unexpected option: --toolset=msvc-142)
When I don't use toolset option i have  error: in triplet x64-windows: Unable to find a valid toolchain for requested target architecture x64.

**Comment by autoantwort** [comment]
You have to create a custom triplet and set https://learn.microsoft.com/en-us/vcpkg/users/triplets#vcpkg_platform_toolset

**Reply by kuzawskak** [reply]
@autoantwort It doesn't work. I got: 
> " error: in triplet x64-windows: Unable to find a valid Visual Studio instance at "C:\Program Files\Microsoft Visual Studio\2022\Professional"
> Could not locate a complete Visual Studio instance"

after setting Env variables like this:

VCPKG_DEFAULT_TRIPLET x64-windows
VCPKG_PLATFORM_TOOLSET v142
VCPKG_PLATFORM_TOOLSET_VERSION 14.31.31103
VCPKG_VISUAL_STUDIO_PATH C:\Program Files\Microsoft Visual Studio\2022\Professional

**Reply by dg0yt** [reply]
> after setting Env variables like this:

Environment variables are not triplet variables.

You are to provide a custom triplet file which is a CMake script. So you would set CMake variables in this script:
~~~cmake
set(VCPKG_PLATFORM_TOOLSET v142)
~~~

**Reply by kuzawskak** [reply]
Well, I created my custom triplet file named x64-windows-custom.cmake
Here it is:
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE dynamic)
set(VCPKG_PLATFORM_TOOLSET v142)
set(VCPKG_PLATFORM_TOOLSET_VERSION 14.31.31103)
set(VCPKG_VISUAL_STUDIO_PATH "C:\\Program Files\\Microsoft Visual Studio\\2022\\Professional")

You can see that the output is the same:

![image](https://github.com/microsoft/vcpkg/assets/6191039/441498f0-067a-4abc-8696-65a961e399bd)

**Reply by kuzawskak** [reply]
Why it doesn't find the compiler?
![image](https://github.com/microsoft/vcpkg/assets/6191039/ce9dce85-8883-4257-b3de-38aad5cdf7ea)

**Reply by dg0yt** [reply]
The error is still for x64-windows. The reason is that x64-windows is still the active host triplet. Override with setting environment variable `VCPKG_DEFAULT_HOST_TRIPLET` to x64-windows-custom, too.
